### PR TITLE
IC-2010: Store and send Delius Office Location Code to CommunityAPI

### PR DESCRIPTION
**Pull request by andrewburrowsmoj**
## What does this pull request do?

Store and send Delius Office Location Code to CommunityAPI when booking or rescheduling an appointment

## What is the intent behind these changes?

Allows to users to specify a Delius office location for an appointment booking

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/validator/AppointmentValidatorTest.kt [hunk 1]
```diff
@@ -79,7 +89,7 @@ internal class AppointmentValidatorTest {
       }
 
       @Test
-      fun `empty address for non nps office appointment throws validation error`() {
+      fun `empty address for a non nps delius appointment throws validation error`() {
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/validator/AppointmentValidatorTest.kt:92]: In the user/domain language we still call this NPS location; can we keep the previous way?

File: src/main/resources/db/migration/V1_68__rename_nps_office_code_column_on_appointment_delivery.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+alter table appointment_delivery rename column nps_office_code to delius_office_location_code;
```
Comment (sldblog) [inline src/main/resources/db/migration/V1_68__rename_nps_office_code_column_on_appointment_delivery.sql:1]: Out of
- `nps_office_code`
- `delius_office_location_code`

the most semantically correct probably would be a third `probation_office_id` -- and using the reference data to make any dereferencing in the translation layer

Context: if we model our domain based on real-life, we're selecting an _office_, not a location code; our context doesn't require location IDs, that becomes relevant when we transition over the boundary to Delius' language

Comment (andrewburrowsmoj) [reply src/main/resources/db/migration/V1_68__rename_nps_office_code_column_on_appointment_delivery.sql:1]: ah so `probation_office_id` is the id on the first column in this csv? https://github.com/ministryofjustice/hmpps-reference-data/blob/main/probation-offices-v0.csv

I personally think it would be better with still using `delius_crs_location_id` for the following reasons:

1. I don't think it's safe to rely on an external data source id since we can't control if this changes. If we had the data out database this would be different since we can add constraints.
2. Similar to 1, I feel it's useful to have a record of the actual code we sent to delius. Yes, this is in reference-data but the connection is weaker and no protected from changes.
3. It means we don't have to implement any reference data integration on services side reducing the complexity for booking an appointment

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentServiceTest.kt [hunk 3]
```diff
@@ -49,6 +49,32 @@ class AppointmentServiceTest {
...
+      durationInMinutes = durationInMinutes,
+      deliusAppointmentId = deliusAppointmentId,
+    )
+    whenever(appointmentRepository.save(any())).thenReturn(savedAppointment)
+
+    // When
+    val newAppointment = appointmentService.createOrUpdateAppointment(referral, null, durationInMinutes, appointmentTime, SUPPLIER_ASSESSMENT, createdByUser, AppointmentDeliveryType.IN_PERSON_MEETING_PROBATION_OFFICE, deliusOfficeLocationCode = deliusOfficeLocationCode)
+
+    // Then
+    verifyResponse(newAppointment, null, true, deliusAppointmentId, appointmentTime, durationInMinutes, AppointmentDeliveryType.PHONE_CALL, deliusOfficeLocationCode)
```
Comment (tomsmyers) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentServiceTest.kt]: why is this still returning PHONE_CALL type?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ActionPlanSessionController.kt [hunk 4]
```diff
@@ -40,7 +40,8 @@ class ActionPlanSessionController(
       updateAppointmentDTO.durationInMinutes,
       user,
       updateAppointmentDTO.appointmentDeliveryType,
-      updateAppointmentDTO.appointmentDeliveryAddress
+      updateAppointmentDTO.appointmentDeliveryAddress,
+      updateAppointmentDTO.deliusOfficeLocationCode,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ActionPlanSessionController.kt:44]: i saw some discussion about the best way to refer to this data - is this what we settled on vs nps_office_code or whatever else was discussed?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/PCCRegionDTO.kt [hunk 5]
```diff
@@ -6,12 +6,14 @@ import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.PCCRegi
 class PCCRegionDTO(
   val id: PCCRegionID,
   val name: String,
+  val npsRegion: NPSRegionDTO,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/PCCRegionDTO.kt:9]: what's the reason for this change?

**Review by tomsmyers — APPROVED** [review]
looks good andrew - thanks

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/PCCRegionDTO.kt (hunk 5 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/PCCRegionDTO.kt:9]: It's in the commit message:

```
IC-2010: Add nps region to pcc region. This is to ensure that the ui is able to understand which nps regions belong to a given pcc region when filtering for office locations.
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ActionPlanSessionController.kt (hunk 4 above)
Comment (andrewburrowsmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ActionPlanSessionController.kt:44]: Sorry it's been a while since I've revisted this PR. I've changed back to the usage of npsOfficeCode, I was originally thinking we'd want to make it more generic.
